chain/chain/src/store.rs

```diff
@@ -2535,61 +2642,476 @@ mod tests {
+        blocks.push((genesis.clone(), Trie::empty_root(), vec![]));
+
+        for simple_chain in chains.iter() {
+            let (source_block, state_root, _) = blocks[simple_chain.from as usize].clone();
+            do_fork(
+                source_block.clone(),
+                state_root,
+                trie.clone(),
+                &mut chain,
+                simple_chain.length,
+                &mut blocks,
+                max_changes,
+                verbose,
+            );
+        }
 
-        assert!(chain.get_block(&blocks[4].hash()).is_ok());
-        assert!(chain.get_block(&blocks[5].hash()).is_err());
-        assert!(chain.get_block(&blocks[6].hash()).is_ok());
-        // block header should be available
-        assert!(chain.get_block_header(&blocks[5].hash()).is_ok());
-        assert!(chain.mut_store().get_all_block_hashes_by_height(5).is_err());
-        assert!(chain.mut_store().get_next_block_hash(&blocks[4].hash()).is_ok());
-        assert!(chain.mut_store().get_next_block_hash(&blocks[5].hash()).is_err());
-        assert!(chain.mut_store().get_next_block_hash(&blocks[6].hash()).is_ok());
         assert!(check_refcount_map(&mut chain).is_ok());
+        // GC execution
+        let clear_data = chain.clear_data(trie.clone());
+        if clear_data.is_err() {
+            println!("clear data failed = {:?}", clear_data);
+            assert!(false);
+        }
+        assert!(check_refcount_map(&mut chain).is_ok());
+
+        // Validation phase
+        let mut chain2 = get_chain();
+        chain2.epoch_length = 10;
+        let trie2 = chain2.runtime_adapter.get_trie().clone();
+
+        // Find gc_height
+        let mut gc_height = chains[0].length - 51;
+        for (i, simple_chain) in chains.iter().enumerate() {
+            if (!simple_chain.is_removed) && gc_height > simple_chain.from && i > 0 {
+                gc_height = simple_chain.from
+            }
+        }
+        if verbose {
+            println!("GC height = {:?}", gc_height);
+        }
+
+        let mut start_index = 1; // zero is for genesis
+        let mut state_roots = vec![Trie::empty_root()];
+        let mut trie_data = HashSet::new();
+
+        for simple_chain in chains {
+            if simple_chain.is_removed {
+                start_index += simple_chain.length;
+                continue;
+            }
+            let (_, mut state_root, _) = blocks[simple_chain.from as usize].clone();
+
+            if trie.iter(&state_root).is_ok() {
+                let a = trie
+                    .iter(&state_root)
+                    .unwrap()
+                    .map(|item| {
+                        if item.is_ok() {
+                            let (key, _) = item.unwrap();
+                            key
+                        } else {
+                            vec![]
+                        }
+                    })
+                    .collect::<Vec<_>>();
+                trie_data.insert(a);
```

Comment:
could you please point the exact places where comments are important?

Reply:
> trie_data is a hashset of vectors of keys?

Correct.

Reply:
> could you please point the exact places where comments are important?

Functions/structs:
- do_fork: ```/// Build a chain of num_blocks on top of prev_block```
- SimpleChain: ```/// from is an index in blocks array, length is the number of blocks in a chain on top of blocks[from], is_removed = ?```

Some places which were hard to understand for me:
- gc_fork_common: what is the test testing? Inputs describe the chain to test on, but what do we check?
- `chain` vs `chain2`: seems like for chain we build the whole tree, and then run gc once?
and for chain2 we apply only blocks which we expect not to be garbage collected?

Reply:
For `trie_data` vs `trie_data2`: trie data will match if root matches and iterator doesn't return errors, so my suggestion is
- run the iterator but `unwrap()` instead of empty vector on error, no need to compare sets. This tests that we don't make trie insconsistent for old blocks / forks
- What we could compare is iterating directly over every `Store` column and making sure two stores match exactly: this way we test that we don't forget to clean anything up

Reply:
@mikhailOK we cannot go forward with your suggestions right now. They are not matching with what's actually going on. Technically, I cannot find the way of comparing two tries honestly.
1. The main issue is GC not remove all state roots in the original Trie1 because some of the GCed nodes are prefixes of the nodes that are alive. Trie1 still keeping them.
2. Cannot unwrap items while iterating.
3. Sets of roots in Trie1 and Trie2 are different.
4. Unclear what behavior of Trie is expected at all.

Reply:
discussed offline